src/coprocessor/memoset/mod.rs

```diff
@@ -0,0 +1,850 @@
+//! transcript as key-value pairs (Lurk data: `(key . value)`); and removals further include the removal multiplicity
+//! (Lurk data: `((key . value) . multiplicity)`). It is critical that the multiplicity be included in the transcript,
+//! since if free to choose it after the randomness has been derived, the prover can trivially falsify the contents of
+//! the multiset -- decoupling claimed truths from those actually proved.
+//!
+//! Bookkeeping required to correctly build the transcript after evaluation but before proving is maintained by the
+//! `Scope`. This allows us to accumulate queries and the subqueries on which they depend, along with the memoized query
+//! results computed 'naturally' during evaluation. We then separate and sort in an order matching that which the NIVC
+//! prover will follow when provably maintaining the multiset accumulator and Fiat-Shamir transcript in the circuit.
+
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use bellpepper_core::{boolean::Boolean, num::AllocatedNum, ConstraintSystem, SynthesisError};
+use itertools::Itertools;
+use once_cell::sync::OnceCell;
+
+use super::gadgets::construct_cons;
+use crate::circuit::gadgets::{
+    constraints::{enforce_equal, enforce_equal_zero, invert, sub},
+    pointer::AllocatedPtr,
+};
+use crate::field::LurkField;
+use crate::lem::circuit::GlobalAllocator;
+use crate::lem::tag::Tag;
+use crate::lem::{pointers::Ptr, store::Store};
+use crate::tag::{ExprTag, Tag as XTag};
+use crate::z_ptr::ZPtr;
+
+use multiset::MultiSet;
+use query::{CircuitQuery, DemoCircuitQuery, Query};
+
+mod multiset;
+mod query;
+
+type ScopeCircuitQuery<F> = DemoCircuitQuery<F>;
+type ScopeQuery<F> = <ScopeCircuitQuery<F> as CircuitQuery<F>>::Q;
+
+#[derive(Clone, Debug)]
+pub struct Transcript<F> {
+    acc: Ptr,
+    _p: PhantomData<F>,
+}
+
+impl<F: LurkField> Transcript<F> {
+    fn new(s: &Store<F>) -> Self {
+        let nil = s.intern_nil();
+        Self {
+            acc: nil,
+            _p: Default::default(),
+        }
+    }
+
+    fn add(&mut self, s: &Store<F>, item: Ptr) {
+        self.acc = s.cons(item, self.acc);
+    }
+
+    fn make_kv(s: &Store<F>, key: Ptr, value: Ptr) -> Ptr {
+        s.cons(key, value)
+    }
+
+    fn make_kv_count(s: &Store<F>, kv: Ptr, count: usize) -> Ptr {
+        let count_num = s.num(F::from_u64(count as u64));
+        s.cons(kv, count_num)
+    }
+
+    /// Since the transcript is just a content-addressed Lurk list, its randomness is the hash value of the associated
+    /// top-level `Cons`. This function sanity-checks the type and extracts that field element.
+    fn r(&self, s: &Store<F>) -> F {
+        let z_ptr = s.hash_ptr(&self.acc);
+        assert_eq!(Tag::Expr(ExprTag::Cons), *z_ptr.tag());
+        *z_ptr.value()
+    }
+
+    #[allow(dead_code)]
+    fn dbg(&self, s: &Store<F>) {
+        //dbg!(self.acc.fmt_to_string_simple(s));
```

Comment:
Oh, this is very much intentional. It's only commented out because our lints won't allow me to leave it in. The method exists specifically for the purpose of providing this `dbg!` output when intended.

Reply:
Others may have workflows that make the tracing version convenient when used in a testing/debugging cycle, but `dbg!` works much better for me and without any fiddling.